ocaml/xapi/helpers.ml

```diff
@@ -2206,3 +2206,138 @@ module BoundedPsq = struct
     let size t = Q.size t.queue
   end
 end
+
+module AuthenticationCache = struct
+  (* Associate arbitrary data with an expiry time. *)
+  module Expires (Data : sig
+    type t
+  end) =
+  struct
+    type t = Data.t with_expiration
+
+    and 'a with_expiration = {data: 'a; expires: Mtime.t}
+
+    let compare {expires= e; _} {expires= e'; _} = Mtime.compare e e'
+  end
+
+  module type Ordered = sig
+    type t
+
+    val compare : t -> t -> int
+  end
+
+  (* A secret associates a digest - derived from a key - with some data. *)
+  module type Secret = sig
+    (** The type of key securing the secret, e.g. a password. *)
+    type key
+
+    (** The type of a digest, derived from a [key] and [salt], e.g. a hashed password. *)
+    type digest
+
+    (** Extra data passed to the hashing routine. *)
+    type salt
+
+    (** The contents of the secret, e.g. an authenticated session. *)
+    type secret
+
+    type t
+
+    val create : digest -> salt -> secret -> t
+
+    val read : t -> digest * salt * secret
+
+    val hash : key -> salt -> digest
+
+    val create_salt : unit -> salt
+
+    val equal_digest : digest -> digest -> bool
+  end
+
+  module type S = sig
+    type t
+
+    type user
+
+    type password
+
+    type session
+
+    val create : size:int -> t
+
+    val cache : t -> user -> password -> session -> unit
+
+    val cached : t -> user -> password -> session option
+  end
+
+  module Make (User : Ordered) (Secret : Secret) :
+    S
+      with type user = User.t
+       and type password = Secret.key
+       and type session = Secret.secret = struct
+    module Q = BoundedPsq.Make (User) (Expires (Secret))
+
+    type user = User.t
+
+    type password = Secret.key
+
+    type session = Secret.secret
+
+    type t = {cache: Q.t; mutex: Mutex.t}
+
+    let create ~size = {cache= Q.create ~capacity:size; mutex= Mutex.create ()}
```

Comment:
This implementation is using Mtime_clock.now (), but it's not used for interprocess communications, I would recommend to use Mtime_clock.counter () and Mtime.Span.t instead:
> Only use [Mtime_clock.now](https://ocaml.org/p/mtime/latest/doc/Mtime_clock/index.html#val-now) if you need inter-process time correlation, otherwise prefer [Mtime_clock.elapsed](https://ocaml.org/p/mtime/latest/doc/Mtime_clock/index.html#val-elapsed) and [counters](https://ocaml.org/p/mtime/latest/doc/Mtime_clock/index.html#counters).

```suggestion
    type t = {cache: Q.t; mutex: Mutex.t; elapsed:Mtime_clock.counter}
    
    let create ~size =
      {cache= Q.create ~capacity:size; mutex= Mutex.create (); elapsed=Mtime_clock.counter ()}
```

Then the keys of the priority queue are differences of time between the creation of the datastructure and the expiry time of the entry. (so it's an Mtime.Span.t). This simplifies the code:

```ocaml
    let cache t user password session =
      let@ () = with_lock t.mutex in
      let expires =
        let elapsed = Mtime_clock.count t.elapsed in
        let timeout = !Xapi_globs.external_authentication_expiry in
        Mtime.Span.add elapsed timeout in
      let salt = Secret.create_salt () in
      let digest = Secret.hash password salt in
      let data = Secret.create digest salt session in
      Q.add t.cache user {data; expires}
```

Comparisons between time spans can be done using the Clock.timer [module](https://github.com/xapi-project/xen-api/blob/42f5d3f06331b7de6244f6ecae30659fce3065a9/ocaml/libs/clock/timer.mli#L43), unfortunately an Mtime version with the [is_longer and is_shorter](https://github.com/dbuenzli/mtime/commit/4118dc79b7bf7b09d157dde7f35f2b8976ab48b6) functions hasn't been released yet.